app/uk/gov/hmrc/gform/testonly/DebugController.scala

```diff
@@ -75,4 +78,217 @@ class DebugController(
+        Future.successful(Ok(page))
+
+    }
+
+  def inspectBooleanExpr(booleanExpr: BooleanExpr, formModel: FormModel[DataExpanded]): BooleanExprInfo = {
+
+    def loop(booleanExpr: BooleanExpr): List[InspectedBooleanExpr] = {
+      def inner(expr1: Expr, expr2: Expr): InspectedBooleanExpr.Binary =
+        InspectedBooleanExpr.Binary(booleanExpr, inspectExpr(expr1, formModel), inspectExpr(expr2, formModel))
+      booleanExpr match {
+        case Equals(left, right)              => inner(left, right) :: Nil
+        case GreaterThan(left, right)         => inner(left, right) :: Nil
+        case GreaterThanOrEquals(left, right) => inner(left, right) :: Nil
+        case LessThan(left, right)            => inner(left, right) :: Nil
+        case LessThanOrEquals(left, right)    => inner(left, right) :: Nil
+        case Not(e)                           => loop(e)
+        case Or(left, right)                  => loop(left) ::: loop(right)
+        case And(left, right)                 => loop(left) ::: loop(right)
+        case IsTrue                           => Nil
+        case IsFalse                          => Nil
+        case Contains(multiValueField, value) =>
+          InspectedBooleanExpr
+            .Contains(booleanExpr, inspectExpr(multiValueField, formModel), inspectExpr(value, formModel)) :: Nil
+        case In(formCtx, dataSource) => Nil
+      }
+    }
+
+    BooleanExprInfo(booleanExpr, loop(booleanExpr))
+  }
+
+  def inspectExpr(expr: Expr, formModel: FormModel[DataExpanded]): InspectedFirstExpr = {
+    val firstOperandInfo: TypeInfo = formModel.toFirstOperandTypeInfo(expr)
+
+    def inner(expr1: Expr, expr2: Expr) = {
+      val typeInfo1 = formModel.toFirstOperandTypeInfo(expr1)
+      val typeInfo2 = formModel.toFirstOperandTypeInfo(expr2)
+      List(typeInfo1, typeInfo2) ::: loop(expr1) ::: loop(expr2)
+    }
+
+    def loop(expr: Expr): List[TypeInfo] =
+      expr match {
+        case Add(expr1: Expr, expr2: Expr)         => inner(expr1, expr2)
+        case Multiply(expr1: Expr, expr2: Expr)    => inner(expr1, expr2)
+        case Subtraction(expr1: Expr, expr2: Expr) => inner(expr1, expr2)
+        case Else(expr1: Expr, expr2: Expr)        => inner(expr1, expr2)
+        case otherwise                             => Nil
+      }
+
+    val typeInfos = NonEmptyList(firstOperandInfo, loop(expr))
+    InspectedFirstExpr(expr, typeInfos)
+  }
+
+  def inspectExplicitExpr(
+    expr: Expr,
+    formComponentId: FormComponentId,
+    formModel: FormModel[DataExpanded]): InspectedExplicitExpr = {
+    val typeInfo: TypeInfo = formModel.explicitTypedExpr(expr, formComponentId)
+
+    def inner(expr1: Expr, expr2: Expr) = {
+      val typeInfo1 = formModel.toFirstOperandTypeInfo(expr1)
+      val typeInfo2 = formModel.toFirstOperandTypeInfo(expr2)
+      List(typeInfo1, typeInfo2) ::: loop(expr1) ::: loop(expr2)
+    }
+
+    def loop(expr: Expr): List[TypeInfo] =
```

Comment:
the inner and loop methods seems identical to the above, could be de-duplicated?

Reply:
Done